README.md
# futuresboard
A python (3.7+) based scraper and dashboard to monitor the performance of your Binance or Bybit Futures account.<br>
<sub>Note: A local sqlite3 database `config/futures.db` will be created and automatically updated by the scraper every 5 minutes.</sub>

[Change log](https://github.com/ecoppen/futuresboard/blob/main/CHANGELOG.md)

[Guide to setting up futuresboard and passivbot](https://www.futuresboard.xyz/)

## Getting started

- Create a fresh new API on Binance or Bybit, with only read rights.
- Clone this repository: `git clone https://github.com/ecoppen/futuresboard.git`
- Navigate to the futuresboard directory: `cd futuresboard`
- Install dependencies `python -m pip install .`. For developing, `python -m pip install -e .[dev]`
- Copy `config/config.json.example` to `config/config.json` and add your new api key and secret: `nano config.json`
- Collect your current trades by running `futuresboard --scrape-only`. If you want to monitor the weight usage (see below).
- By default, when launching the futuresboard web application, a separate thread is also started to continuously collect new trades.
  Alternatively, setup the scraper on a crontab or alternative to keep the database up-to-date: `crontab -e` then `*/5 * * * * futuresboard --scrape-only` (example is every 5 minutes, change to your needs)
  In this case, don't forget to pass `--disable-auto-scraper`.
- Start a screen or alternative if you want the webserver to persist: `screen -S futuresboard`
- Start the futuresboard web application `futuresboard`
- Navigate to the IP address shown e.g. `http://127.0.0.1:5000/`. These settings can be changed by passing `--host` and/or `--port` when running the above command

Currently only Binance and Bybit Futures are supported - as those are supported by passivbot.

## API weight usage - Binance

- Reminder: Binance API allows you to consume up to `1200 weight / minute / IP`.
- Account: Fetching account information costs `5` weight per run
- Income: Fetching income information costs `30` weight per 1000 (initial run will build database, afterwards only new income will be fetched)
- Orders: Fetching open order information costs `40` weight per run
- The scraper will sleep for a minute when the rate exceeds `800 within a minute`

## API weight usage - Bybit
- Account/Income/Positions: Fetching account information or positions costs `1` weight per run with a maximum combination allowed of `120/min`. Income can be fetched in batches of 50 (initial run will build database, afters only new income will be fetched)
- Orders: Fetching open order information costs `1` weight per symbol with a maximum allowed of `600/min`
- The scraper will sleep for a minute when the rate exceeds `100 within a minute`

## Customising the dashboard
The `/config/config.json` file allows you to customise the look and feel of your dashboard as follows:
- `AUTO_SCRAPE_INTERVAL` is set to 300 seconds, this value can be adjusted between `60` and `3600`
- `NAVBAR_TITLE` changes the branding in the top left of the navigation (see below)
- `NAVBAR_BG` changes the colour of the navigation bar, acceptable values are: `bg-primary`, `bg-secondary`, `bg-success`, `bg-danger`, `bg-warning`, `bg-info` and the default `bg-dark`
- `PROJECTIONS` changes the percentage values on the projections page. `1.003` equates to `0.3%` daily and `1.01` equates to `1%` daily.

For example, setting `"NAVBAR_TITLE": "Custom title"` and `"NAVBAR_BG": "bg-primary",` would result in:
<img width="1314" src="https://user-images.githubusercontent.com/51025241/145480528-408dff64-1742-41ea-baac-89bb5458d406.png">
<img width="500" src="https://user-images.githubusercontent.com/51025241/145609351-631db009-ac04-47c9-ae82-0d76af0362d2.png">
## Screenshots
<img width="600" alt="dashboard" src="https://user-images.githubusercontent.com/51025241/147236951-c87d1b2e-9eee-49bb-bc44-1769b9756f45.png">
<img width="600" alt="calendar" src="https://user-images.githubusercontent.com/51025241/147236947-426ee144-fe30-4041-93b0-36a3073a9233.png">
<img width="550" alt="coin" src="https://user-images.githubusercontent.com/51025241/147359139-e2ba33c9-0d10-4b09-a235-c9979658dd9b.png">
<img width="600" alt="1h chart" src="https://user-images.githubusercontent.com/51025241/147358812-b860f5db-a867-4bf4-b98e-b467bb928a25.png">
<img width="600" alt="history" src="https://user-images.githubusercontent.com/51025241/147236956-7b427c72-0a8b-443b-bd24-3eaef3246895.png">
<img width="600" alt="positions" src="https://user-images.githubusercontent.com/51025241/147236958-160a4cd8-c461-46d2-87d1-560d89207a93.png">
<img width="600" alt="projection" src="https://user-images.githubusercontent.com/51025241/147236959-7ca52391-f6bb-496e-bba2-5b914ee333c7.png">

## Alternative dashboards

- https://github.com/hoeckxer/exchanges_dashboard
- https://github.com/SH-Stark/trading-dashboard


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)

## [Unreleased]
- Match all of the details from [passivbot](https://github.com/enarjord/passivbot)s telegram commands `/position` and `/open_orders` commands
- Store and display historical unrealized PnL
- Tidy up repeated code into functions

## 2022-01-03
### Changed
- Added a composite key to the income table in the database to stop duplicates from being entered. This change will only be active on fresh databases/built from scratch. 
- Added the transactionId from Bybit to the list for future processing - this is a string whereas Binance uses integers. This needs to be managed to avoid more database deletions.

## 2021-12-27
### Changed
- Fixed an issue on history page where dowload buttons would cause 500 error when pressed
- Fixed an issue with bybit scraping to get latest trades. Still more to do here with regard to old trades looking back.

## 2021-12-25
### Changed
- Commas removed from `config.json.example` as the extra commas caused errors [@jontstaz](https://github.com/jontstaz)
- Inverse the colours for current buy/sells on the 15m, 1h, 4h and 1d charts to match green for buy and red for sell

## 2021-12-24
### Added
- Bybit now supported as an exchange, data will be scraped as best as possible however whole historical data doesn't seem possible currently.
- Candle chart with buy/sell lines now available on coin pages for 15m, 1h, 4h and 1d

### Changed
- Coin pages now have Average Down Calculator showing all the time, with specific distances away from position already calculated
- Income history page buttons fixed
- Positions page now has a table showing distance from next buys and all previous information

## 2021-12-10
### Changed
- Minor fixes and README updates regarding the latest changes made yesterday

## 2021-12-09
### Added
- The config file now has the ability to change the look and feel of the dashboard. The title can be changed using `NAVBAR_TITLE`, the background colour of the navigation bar can now be changed to a set of colours using `NAVBAR_BG` and the projection percentages can be changed to show more or less values as specified by the user with `PROJECTIONS`
- The config file can also be used to set the value of the `AUTO_SCRAPE_INTERVAL` between `60` and `3600` for more or less frequent updates depending on their needs
- Under the charts on the dashboard and coin pages there is now a summary showing either the `total profit` for that period or `number of coins active`
- Positions page now shows unrealised Profit and Loss for each coin

### Changed
- Tooltips were lost somewhere along the line for the values in the sidebar, these are reintroduced 
- Link to the repo in the footer was lost also, reintroduced

## 2021-12-07
### Changed
- Codebase overhauled by [@s0undt3ch](https://github.com/s0undt3ch) to use a blueprint which allows the multiple dashboards to be served 
- Hard coded URLs swapped to use `url_for` by  [@s0undt3ch](https://github.com/s0undt3ch)

## 2021-12-02
### Changed
- Timeframes no longer exist, instead there are start dates and end dates with a daterangepicker on all appropriate pages

## 2021-11-29
### Changed
- Added a fix for the graph on the dashboard that would only calculate the total - profit over a week and not the other tiemframes

## 2021-11-28
### Added
- Route added for balance projection over 365 days at fixed and average %, this can be accessed from the main dashboard page above the line chart
- Average down calculator now available on coin page (only shows when a coin has no more buy positions)

### Changed
- Changed 'Week' to be previous 7 days rather than 8

## 2021-11-21
### Changed
- Fixed issue that was reintroduced (new positions wouldn't show unless a buy/sell was present)
- Fixed issue where `positions` were not returned by API causing error which is now handled more gracefully
- Fixed issue where some coins were automatically hidden on the graph when too many were present - now resizes to fit

## 2021-11-20
### Added
- Table on history page to show all of the previously saved CSV files
- favicon2 (warning emoji) will replace favicon (rocket emoji) when no buys are left on an active coin
- New route added for `All positions` that shows every coin in one place. Accordian can be opened to see the orders relating to each position.

### Changed
- Active coins now have a red background if there are no buys left on an active coin
- Cards on the front page (profit for today, week, month and PnL) are now hyperlinked to their respective pages


src/futuresboard/scraper.py
# https://github.com/binance/binance-signature-examples
from __future__ import annotations

import datetime
import hashlib
import hmac
import sqlite3
import threading
import time
from datetime import timedelta
from sqlite3 import Error
from urllib.parse import urlencode
from collections import OrderedDict

import requests
from flask import current_app


class HTTPRequestError(Exception):
    def __init__(self, url, code, msg=None):
        self.url = url
        self.code = code
        self.msg = msg

    def __str__(self) -> str:
        """
        Convert the exception into a printable string
        """
        return f"Request to {self.url!r} failed. Code: {self.code}; Message: {self.msg}"


def auto_scrape(app):
    thread = threading.Thread(target=_auto_scrape, args=(app,))
    thread.daemon = True
    thread.start()


def _auto_scrape(app):
    with app.app_context():
        interval = app.config["AUTO_SCRAPE_INTERVAL"]
        while True:
            app.logger.info("Auto scrape routines starting")
            scrape(app=app)
            app.logger.info("Auto scrape routines terminated. Sleeping %s seconds...", interval)
            time.sleep(interval)


def hashing(query_string):
    return hmac.new(
        current_app.config["API_SECRET"].encode("utf-8"),
        query_string.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()


def get_timestamp():
    return int(time.time() * 1000)


def dispatch_request(http_method):
    session = requests.Session()
    session.headers.update(
        {
            "Content-Type": "application/json;charset=utf-8",
            "X-MBX-APIKEY": current_app.config["API_KEY"],
        }
    )
    return {
        "GET": session.get,
        "DELETE": session.delete,
        "PUT": session.put,
        "POST": session.post,
    }.get(http_method, "GET")


# used for sending request requires the signature
def send_signed_request(http_method, url_path, payload={}, signature="signature"):
    payload["timestamp"] = get_timestamp()
    query_string = urlencode(OrderedDict(sorted(payload.items())))
    query_string = query_string.replace("%27", "%22")  # replace single quote to double quote

    url = (
        current_app.config["API_BASE_URL"]
        + url_path
        + "?"
        + query_string
        + "&"
        + signature
        + "="
        + hashing(query_string)
    )

    # print("{} {}".format(http_method, url))
    params = {"url": url, "params": {}}
    response = dispatch_request(http_method)(**params)
    headers = response.headers
    json_response = response.json()
    if "code" in json_response:
        raise HTTPRequestError(url=url, code=json_response["code"], msg=json_response["msg"])
    return headers, json_response


# used for sending public data request
def send_public_request(url_path, payload={}):
    query_string = urlencode(payload, True)
    url = current_app.config["API_BASE_URL"] + url_path
    if query_string:
        url = url + "?" + query_string
    # print("{}".format(url))
    response = dispatch_request("GET")(url=url)
    headers = response.headers
    json_response = response.json()
    if "code" in json_response:
        raise HTTPRequestError(url=url, code=json_response["code"], msg=json_response["msg"])
    return headers, json_response


def create_connection(db_file):
    conn = None
    try:
        conn = sqlite3.connect(db_file)
    except Error as e:
        print(e)

    return conn


def create_table(conn, create_table_sql):
    try:
        c = conn.cursor()
        c.execute(create_table_sql)
    except Error as e:
        print(e)


def db_setup(database):
    sql_create_income_table = """ CREATE TABLE IF NOT EXISTS income (
                                        IID integer PRIMARY KEY AUTOINCREMENT,
                                        tranId integer,
                                        symbol text,
                                        incomeType text,
                                        income real,
                                        asset text,
                                        info text,
                                        time integer,
                                        tradeId integer,
                                        UNIQUE(tranId, incomeType) ON CONFLICT REPLACE
                                    ); """

    sql_create_position_table = """ CREATE TABLE IF NOT EXISTS positions (
                                        PID integer PRIMARY KEY AUTOINCREMENT,
                                        symbol text,
                                        unrealizedProfit real,
                                        leverage integer,
                                        entryPrice real,
                                        positionSide text,
                                        positionAmt real
                                    ); """

    sql_create_account_table = """ CREATE TABLE IF NOT EXISTS account (
                                        AID integer PRIMARY KEY,
                                        totalWalletBalance real,
                                        totalUnrealizedProfit real,
                                        totalMarginBalance real,
                                        availableBalance real,
                                        maxWithdrawAmount real
                                    ); """

    sql_create_orders_table = """ CREATE TABLE IF NOT EXISTS orders (
                                        OID integer PRIMARY KEY AUTOINCREMENT,
                                        origQty real,
                                        price real,
                                        side text,
                                        positionSide text,
                                        status text,
                                        symbol text,
                                        time integer,
                                        type text
                                    ); """
    # create a database connection
    conn = create_connection(database)

    # create tables
    if conn is not None:
        create_table(conn, sql_create_income_table)
        create_table(conn, sql_create_position_table)
        create_table(conn, sql_create_account_table)
        create_table(conn, sql_create_orders_table)
    else:
        print("Error! cannot create the database connection.")


# income interactions
def create_income(conn, income):
    sql = """ INSERT INTO income(tranId, symbol, incomeType, income, asset, info, time, tradeId)
              VALUES(?,?,?,?,?,?,?,?) """
    cur = conn.cursor()
    cur.execute(sql, income)


def select_latest_income(conn):
    cur = conn.cursor()
    cur.execute("SELECT time FROM income ORDER BY time DESC LIMIT 0, 1")
    return cur.fetchone()


def select_latest_income_symbol(conn, symbol):
    cur = conn.cursor()
    cur.execute("SELECT time FROM income WHERE symbol = ? ORDER BY time DESC LIMIT 0, 1", (symbol,))
    return cur.fetchone()


# position interactions
def create_position(conn, position):
    sql = """ INSERT INTO positions(unrealizedProfit, leverage, entryPrice, positionAmt, symbol, positionSide) VALUES(?,?,?,?,?,?) """
    cur = conn.cursor()
    cur.execute(sql, position)


def update_position(conn, position):
    sql = """ UPDATE positions SET unrealizedProfit = ?, leverage = ?, entryPrice = ?, positionAmt = ? WHERE symbol = ? AND positionSide = ? """
    cur = conn.cursor()
    cur.execute(sql, position)


def select_position(conn, symbol):
    cur = conn.cursor()
    cur.execute("SELECT unrealizedProfit FROM positions WHERE symbol = ? LIMIT 0, 1", (symbol,))
    return cur.fetchone()


# account interactions
def create_account(conn, account):
    sql = """ INSERT INTO account(totalWalletBalance, totalUnrealizedProfit, totalMarginBalance, availableBalance, maxWithdrawAmount, AID) VALUES(?,?,?,?,?,?) """
    cur = conn.cursor()
    cur.execute(sql, account)


def update_account(conn, account):
    sql = """ UPDATE account SET totalWalletBalance = ?, totalUnrealizedProfit = ?, totalMarginBalance = ?, availableBalance = ?, maxWithdrawAmount = ? WHERE AID = ?"""
    cur = conn.cursor()
    cur.execute(sql, account)


def select_account(conn):
    cur = conn.cursor()
    cur.execute("SELECT totalWalletBalance FROM account LIMIT 0, 1")
    return cur.fetchone()


# orders interactions
def delete_all_orders(conn):
    sql = """ DELETE FROM orders """
    cur = conn.cursor()
    cur.execute(sql)
    conn.commit()


def create_orders(conn, orders):
    sql = """ INSERT INTO orders(origQty, price, side, positionSide, status, symbol, time, type) VALUES(?,?,?,?,?,?,?,?) """
    cur = conn.cursor()
    cur.execute(sql, orders)


def scrape(app=None):
    try:
        _scrape(app=app)
    except HTTPRequestError as exc:
        if app is None:
            print(exc)
        else:
            app.logger.error(str(exc))


def _scrape(app=None):
    start = time.time()
    db_setup(current_app.config["DATABASE"])

    up_to_date = False
    weightused = 0
    processed, updated_positions, new_positions, updated_orders, sleeps = 0, 0, 0, 0, 0

    if current_app.config["EXCHANGE"].lower() == "binance":
        if weightused < 800:
            responseHeader, responseJSON = send_signed_request("GET", "/fapi/v1/openOrders")
            weightused = int(responseHeader["X-MBX-USED-WEIGHT-1M"])

            with create_connection(current_app.config["DATABASE"]) as conn:
                delete_all_orders(conn)
                for order in responseJSON:
                    updated_orders += 1
                    row = (
                        float(order["origQty"]),
                        float(order["price"]),
                        order["side"],
                        order["positionSide"],
                        order["status"],
                        order["symbol"],
                        int(order["time"]),
                        order["type"],
                    )
                    create_orders(conn, row)
                conn.commit()

        responseHeader, responseJSON = send_signed_request("GET", "/fapi/v2/account")
        weightused = int(responseHeader["X-MBX-USED-WEIGHT-1M"])

        overweight = False
        try:
            positions = responseJSON["positions"]
        except Exception:
            overweight = True

        if not overweight:
            with create_connection(current_app.config["DATABASE"]) as conn:
                totals_row = (
                    float(responseJSON["totalWalletBalance"]),
                    float(responseJSON["totalUnrealizedProfit"]),
                    float(responseJSON["totalMarginBalance"]),
                    float(responseJSON["availableBalance"]),
                    float(responseJSON["maxWithdrawAmount"]),
                    1,
                )
                accountCheck = select_account(conn)
                if accountCheck is None:
                    create_account(conn, totals_row)
                elif float(accountCheck[0]) != float(responseJSON["totalWalletBalance"]):
                    update_account(conn, totals_row)

                for position in positions:
                    position_row = (
                        float(position["unrealizedProfit"]),
                        int(position["leverage"]),
                        float(position["entryPrice"]),
                        float(position["positionAmt"]),
                        position["symbol"],
                        position["positionSide"],
                    )
                    unrealizedProfit = select_position(conn, position["symbol"])
                    if unrealizedProfit is None:
                        create_position(conn, position_row)
                        new_positions += 1
                    elif float(unrealizedProfit[0]) != float(position["unrealizedProfit"]):
                        update_position(conn, position_row)
                        updated_positions += 1

                conn.commit()

        while not up_to_date:
            if weightused > 800:
                print(f"Weight used: {weightused}/800\nProcessed: {processed}\nSleep: 1 minute")
                sleeps += 1
                time.sleep(60)

            with create_connection(current_app.config["DATABASE"]) as conn:
                startTime = select_latest_income(conn)
                if startTime is None:
                    startTime = int(
                        datetime.datetime.fromisoformat("2020-01-01 00:00:00+00:00").timestamp()
                        * 1000
                    )
                else:
                    startTime = startTime[0]

                params = {"startTime": startTime + 1, "limit": 1000}

                responseHeader, responseJSON = send_signed_request("GET", "/fapi/v1/income", params)
                weightused = int(responseHeader["X-MBX-USED-WEIGHT-1M"])

                if len(responseJSON) == 0:
                    up_to_date = True
                else:
                    for income in responseJSON:
                        if len(income["tradeId"]) == 0:
                            income["tradeId"] = 0
                        income_row = (
                            int(income["tranId"]),
                            income["symbol"],
                            income["incomeType"],
                            income["income"],
                            income["asset"],
                            income["info"],
                            int(income["time"]),
                            int(income["tradeId"]),
                        )
                        create_income(conn, income_row)
                        processed += 1

                    conn.commit()
    elif current_app.config["EXCHANGE"].lower() == "bybit":
        all_symbols = []
        exec_type = {
            "Trade": "REALIZED_PNL",
            "Funding": "FUNDING_FEE",
            "AdlTrade": "ADLTRADE",
            "BustTrade": "BUSTTRADE",
        }

        params = {"api_key": current_app.config["API_KEY"]}
        responseHeader, responseJSON = send_signed_request(
            "GET", "/private/linear/position/list", params, signature="sign"
        )
        weightused = int(responseJSON["rate_limit_status"])

        with create_connection(current_app.config["DATABASE"]) as conn:
            delete_all_orders(conn)

            for position in responseJSON["result"]:
                if weightused < 10:
                    print(
                        f"Weight used: {weightused}/{120-weightused}\nProcessed: {updated_positions + new_positions + updated_orders}\nSleep: 1 minute"
                    )
                    sleeps += 1
                    time.sleep(60)

                if position["data"]["symbol"] not in all_symbols:
                    all_symbols.append(position["data"]["symbol"])

                if position["data"]["size"] > 0:
                    if position["data"]["side"].lower() == "buy":
                        positionside = "LONG"
                    else:
                        positionside = "SHORT"

                    position_row = (
                        float(position["data"]["unrealised_pnl"]),
                        int(position["data"]["leverage"]),
                        float(position["data"]["entry_price"]),
                        float(position["data"]["size"]),
                        position["data"]["symbol"],
                        positionside,
                    )
                    unrealizedProfit = select_position(conn, position["data"]["symbol"])
                    if unrealizedProfit is None:
                        create_position(conn, position_row)
                        new_positions += 1
                    elif float(unrealizedProfit[0]) != float(position["data"]["unrealised_pnl"]):
                        update_position(conn, position_row)
                        updated_positions += 1

                    params = {
                        "symbol": position["data"]["symbol"],
                        "api_key": current_app.config["API_KEY"],
                    }
                    responseHeader, responseJSON = send_signed_request(
                        "GET", "/private/linear/order/search", params, signature="sign"
                    )
                    weightused = int(responseJSON["rate_limit_status"])

                    for order in responseJSON["result"]:

                        updated_orders += 1

                        if order["side"].lower() == "buy":
                            orderside = "BUY"
                        else:
                            orderside = "SELL"

                        time_format = datetime.datetime.strptime(
                            order["created_time"], "%Y-%m-%dT%H:%M:%SZ"
                        )

                        row = (
                            float(order["qty"]),
                            float(order["price"]),
                            orderside,
                            positionside,
                            order["order_status"],
                            order["symbol"],
                            int(time_format.timestamp() * 1000),
                            order["order_type"],
                        )
                        create_orders(conn, row)

            params = {"api_key": current_app.config["API_KEY"], "coin": "USDT"}
            responseHeader, responseJSON = send_signed_request(
                "GET", "/v2/private/wallet/balance", params, signature="sign"
            )

            totals_row = (
                float(responseJSON["result"]["USDT"]["wallet_balance"]),
                float(responseJSON["result"]["USDT"]["unrealised_pnl"]),
                float(responseJSON["result"]["USDT"]["used_margin"]),
                float(responseJSON["result"]["USDT"]["available_balance"]),
                float(0),
                1,
            )

            accountCheck = select_account(conn)
            if accountCheck is None:
                create_account(conn, totals_row)
            elif float(accountCheck[0]) != float(responseJSON["result"]["USDT"]["wallet_balance"]):
                update_account(conn, totals_row)

            conn.commit()

        all_symbols = sorted(all_symbols)

        for symbol in all_symbols:
            trades = {}
            params = {"api_key": current_app.config["API_KEY"], "symbol": symbol, "limit": 50}
            with create_connection(current_app.config["DATABASE"]) as conn:
                startTime = select_latest_income_symbol(conn, symbol)
                if startTime is None:
                    startTime = int(
                        datetime.datetime.fromisoformat("2020-01-01 00:00:00+00:00").timestamp()
                    )
                    params["start_time"] = startTime
                else:
                    startTime = int(startTime[0]) / 1000
                    params["start_time"] = int(startTime + 1)

            for page in range(1, 50):
                if weightused < 20:
                    print(f"Weight used: {weightused}/100\nProcessed: {processed}\nSleep: 1 minute")
                    sleeps += 1
                    time.sleep(60)
                params["page"] = page
                responseHeader, responseJSON = send_signed_request(
                    "GET", "/private/linear/trade/closed-pnl/list", params, signature="sign"
                )
                if "rate_limit_status" in responseJSON:
                    weightused = int(responseJSON["rate_limit_status"])

                if responseJSON["result"] is not None:
                    if responseJSON["result"]["data"] is not None:
                        for trade in responseJSON["result"]["data"]:
                            trades[trade["created_at"]] = [
                                trade["id"],
                                trade["exec_type"],
                                trade["closed_pnl"],
                                trade["order_id"],
                            ]
                        if len(responseJSON["result"]["data"]) < 50:
                            break
                    else:
                        break
                else:
                    break

            if len(trades) > 0:
                trades = OrderedDict(sorted(trades.items()))
                with create_connection(current_app.config["DATABASE"]) as conn:
                    for trade in trades:
                        income_row = (
                            int(trades[trade][0]),
                            symbol,
                            exec_type[trades[trade][1]],
                            trades[trade][2],
                            "USDT",
                            exec_type[trades[trade][1]],
                            int(trade * 1000),
                            int(trades[trade][0]),
                        )

                        create_income(conn, income_row)
                        processed += 1
                    conn.commit()
    else:
        current_app.logger.info(
            "Exchange; %s is not currently supported", current_app.config["EXCHANGE"]
        )

    elapsed = time.time() - start
    if app is not None:
        current_app.logger.info(
            "Orders updated: %s; Positions updated: %s (new: %s); Trades processed: %s; Time elapsed: %s; Sleeps: %s",
            updated_orders,
            updated_positions,
            new_positions,
            processed,
            timedelta(seconds=elapsed),
            sleeps,
        )
    else:
        print(
            "Orders updated: {}\nPositions updated: {} (new: {})\nTrades processed: {}\nTime elapsed: {}\nSleeps: {}".format(
                updated_orders,
                updated_positions,
                new_positions,
                processed,
                timedelta(seconds=elapsed),
                sleeps,
            )
        )
